File: src/MeisterDev.ProPR.Application/Features/UsageStatistics/Services/UsageStatisticsService.cs
// Copyright (c) Andreas Rain.
// Licensed under the Elastic License 2.0. See LICENSE file in the project root for full license terms.

using MeisterDev.ProPR.Application.Features.UsageStatistics.Dtos;
using MeisterDev.ProPR.Application.Features.UsageStatistics.Models;
using MeisterDev.ProPR.Application.Features.UsageStatistics.Ports;
using MeisterDev.ProPR.Application.Features.UsageStatistics.Support;
using MeisterDev.ProPR.Application.Interfaces;
using MeisterDev.ProPR.Application.Support;

namespace MeisterDev.ProPR.Application.Features.UsageStatistics.Services;

/// <summary>
///     The operations the administration UI performs on anonymous usage statistics.
///     <para>
///         This class has no transport dependency. Reading the settings, previewing the payload and dismissing
///         the notice are local operations, so the payload preview works in an installation that has never sent
///         anything.
///     </para>
/// </summary>
public sealed class UsageStatisticsService(
    IUsageStatisticsStateStore stateStore,
    UsageStatisticsSnapshotBuilder snapshotBuilder,
    UsageStatisticsEditionResolver editionResolver,
    IProductVersionProvider productVersionProvider,
    UsageStatisticsSender? sender = null)
{
    /// <summary>The message shown when a commercial license governs the control.</summary>
    public const string ManagedByLicenseMessage =
        "Anonymous usage statistics are managed by your commercial license.";

    /// <summary>Returns the current settings, including the last send outcome and any update information.</summary>
    public async Task<UsageStatisticsSettingsDto> GetSettingsAsync(CancellationToken cancellationToken = default)
    {
        var state = await stateStore.GetAsync(cancellationToken);
        var edition = await editionResolver.ResolveAsync(cancellationToken);

        return this.ToDto(state, edition);
    }

    /// <summary>
    ///     Stores the community toggle.
    ///     <para>
    ///         Refused while a license is installed. The control stays visible in that state rather than being
    ///         hidden, so administrators can still see what the installation sends.
    ///     </para>
    /// </summary>
    /// <exception cref="InvalidOperationException">A commercial license governs the setting.</exception>
    public async Task<UsageStatisticsSettingsDto> SetCommunityOptInAsync(
        bool optIn,
        Guid? actorUserId,
        CancellationToken cancellationToken = default)
    {
        var edition = await editionResolver.ResolveAsync(cancellationToken);
        if (edition == UsageStatisticsEdition.Commercial)
        {
            throw new InvalidOperationException(ManagedByLicenseMessage);
        }

        var state = await stateStore.SetCommunityOptInAsync(optIn, actorUserId, cancellationToken);
        return this.ToDto(state, edition);
    }

    /// <summary>
    ///     Records that the consent notice was shown to an administrator, which opens the gate in community
    ///     installations.
    /// </summary>
    public async Task<UsageStatisticsSettingsDto> RecordNoticeShownAsync(CancellationToken cancellationToken = default)
    {
        var state = await stateStore.RecordConsentGateSatisfiedAsync(cancellationToken);
        var edition = await editionResolver.ResolveAsync(cancellationToken);

        return this.ToDto(state, edition);
    }

    /// <summary>Hides the notice for this installation. Dismissal changes nothing about what is sent.</summary>
    public async Task<UsageStatisticsSettingsDto> DismissNoticeAsync(CancellationToken cancellationToken = default)
    {
        var state = await stateStore.RecordNoticeDismissedAsync(cancellationToken);
        var edition = await editionResolver.ResolveAsync(cancellationToken);

        return this.ToDto(state, edition);
    }

    /// <summary>
    ///     Opens the gate on an administrator's sign-in, for installations that show no consent notice.
    ///     <para>
    ///         Community installations are gated on the notice rendering instead, which requires the
    ///         explanation to have been displayed rather than only that an administrator signed in.
    ///     </para>
    /// </summary>
    public async Task RecordAdministratorSignInAsync(CancellationToken cancellationToken = default)
    {
        var edition = await editionResolver.ResolveAsync(cancellationToken);
        if (edition != UsageStatisticsEdition.Commercial)
        {
            return;
        }

        var state = await stateStore.GetAsync(cancellationToken);
        if (state.IsConsentGateSatisfied)
        {
            return;
        }

        await stateStore.RecordConsentGateSatisfiedAsync(cancellationToken);
    }

    /// <summary>
    ///     Runs a send cycle now instead of waiting for the daily one.
    ///     <para>
    ///         This runs the same cycle as the background loop, under the same rules: an installation that is
    ///         off or has not shown the notice sends nothing, and one that already sent today returns not due
    ///         rather than sending a second snapshot. The documented limit of one snapshot a day still applies.
    ///     </para>
    /// </summary>
    /// <exception cref="InvalidOperationException">The installation cannot send at all.</exception>
    public async Task<UsageStatisticsSendResultDto> SendNowAsync(CancellationToken cancellationToken = default)
    {
        if (sender is null)
        {
            throw new InvalidOperationException("This installation has no usage statistics sender configured.");
        }

        var result = await sender.SendIfDueAsync(cancellationToken);

        return new UsageStatisticsSendResultDto(result.Decision, await this.GetSettingsAsync(cancellationToken));
    }

    /// <summary>Builds the payload the next ping would carry, without sending anything.</summary>
    public async Task<UsageStatisticsPreviewDto> BuildPreviewAsync(CancellationToken cancellationToken = default)
    {
        var state = await stateStore.GetAsync(cancellationToken);
        var edition = await editionResolver.ResolveAsync(cancellationToken);
        var snapshot = await snapshotBuilder.BuildAsync(state, edition, cancellationToken);

        return new UsageStatisticsPreviewDto(
            UsageStatisticsContract.PingEndpoint,
            "application/json",
            UsageStatisticsContract.Serialize(snapshot),
            UsageStatisticsContract.PayloadDocumentationUrl);
    }

    private UsageStatisticsSettingsDto ToDto(UsageStatisticsState state, UsageStatisticsEdition edition)
    {
        var managedByLicense = edition == UsageStatisticsEdition.Commercial;

        return new UsageStatisticsSettingsDto(
            edition,
            state.IsSendingEnabled(edition),
            state.CommunityOptIn,
            managedByLicense,
            state.IsConsentGateSatisfied,
            !managedByLicense && state.NoticeDismissedAt is null,
            state.LastAttemptAt,
            state.LastAttemptSucceeded,
            state.LastAttemptDetail,
            state.LastSuccessAt,
            UsageStatisticsContract.PingEndpoint,
            UsageStatisticsContract.PayloadDocumentationUrl,
            UsageStatisticsContract.PrivacyContact,
            this.ToUpdateStatus(state));
    }

    private UsageStatisticsUpdateStatusDto ToUpdateStatus(UsageStatisticsState state)
    {
        var currentVersion = productVersionProvider.Version;

        // An unstamped development build has no release number to compare against, so it never reports an
        // available update. Every local build would otherwise show the update badge.
        var comparable = !string.Equals(
            currentVersion,
            AssemblyProductVersionProvider.UnstampedVersion,
            StringComparison.Ordinal);

        var updateAvailable = comparable
                              && !string.IsNullOrWhiteSpace(state.LatestVersion)
                              && !string.Equals(state.LatestVersion, currentVersion, StringComparison.OrdinalIgnoreCase);

        return new UsageStatisticsUpdateStatusDto(
            currentVersion,
            state.LatestVersion,
            updateAvailable,
            state.Advisories,
            state.UpdateInformationReceivedAt);
    }
}

File: frontend/src/services/usageStatisticsService.ts
// Copyright (c) Andreas Rain.
// Licensed under the Elastic License 2.0. See LICENSE file in the project root for full license terms.

import { createAdminClient, getApiErrorMessage } from '@/services/api'
import type { components } from '@/types'

/**
 * Hand-written response shapes, as the other service wrappers do.
 *
 * The generated schema marks every property optional, because the API's OpenAPI document does not emit
 * `required`. These endpoints always answer with the whole object, so restating the shape here keeps the
 * optional chaining out of three components.
 */
export type UsageStatisticsEdition = components['schemas']['UsageStatisticsEdition']

export interface ProductAdvisory {
  id: string
  severity: string
  title?: string | null
  affectedVersions?: string | null
  link?: string | null
}

export interface UsageStatisticsUpdateStatus {
  currentVersion: string
  latestVersion: string | null
  updateAvailable: boolean
  advisories: ProductAdvisory[]
  receivedAt: string | null
}

export interface UsageStatisticsSettings {
  edition: UsageStatisticsEdition
  enabled: boolean
  communityOptIn: boolean
  managedByLicense: boolean
  consentGateSatisfied: boolean
  noticeRequired: boolean
  lastAttemptAt: string | null
  lastAttemptSucceeded: boolean | null
  lastAttemptDetail: string | null
  lastSuccessAt: string | null
  pingEndpoint: string
  payloadDocumentationUrl: string
  privacyContact: string
  update: UsageStatisticsUpdateStatus
}

export interface UsageStatisticsPreview {
  endpoint: string
  contentType: string
  payload: string
  payloadDocumentationUrl: string
}

export type UsageStatisticsSendDecision = components['schemas']['UsageStatisticsSendDecision']

export interface UsageStatisticsSendResult {
  decision: UsageStatisticsSendDecision
  settings: UsageStatisticsSettings
}

function getClient() {
  return createAdminClient()
}

export async function getUsageStatisticsSettings(): Promise<UsageStatisticsSettings> {
  const { data, error, response } = await getClient().GET('/admin/usage-statistics', {})

  if (!response.ok) {
    throw new Error(getApiErrorMessage(error, 'Failed to load the anonymous usage statistics setting.'))
  }

  return data as UsageStatisticsSettings
}

export async function setUsageStatisticsEnabled(enabled: boolean): Promise<UsageStatisticsSettings> {
  const { data, error, response } = await getClient().PATCH('/admin/usage-statistics', {
    body: { enabled },
  })

  if (!response.ok) {
    throw new Error(getApiErrorMessage(error, 'Failed to update the anonymous usage statistics setting.'))
  }

  return data as UsageStatisticsSettings
}

/**
 * Runs a send cycle now instead of waiting for the daily one.
 *
 * Every rule the daily loop applies still applies, so this cannot send more than the documented one snapshot
 * a day. The returned decision distinguishes the cases in which nothing was sent, which the settings alone do
 * not.
 */
export async function sendUsageStatisticsNow(): Promise<UsageStatisticsSendResult> {
  const { data, error, response } = await getClient().POST('/admin/usage-statistics/send', {})

  if (!response.ok) {
    throw new Error(getApiErrorMessage(error, 'The snapshot could not be sent.'))
  }

  return data as UsageStatisticsSendResult
}

export async function getUsageStatisticsPreview(): Promise<UsageStatisticsPreview> {
  const { data, error, response } = await getClient().GET('/admin/usage-statistics/preview', {})

  if (!response.ok) {
    throw new Error(getApiErrorMessage(error, 'Failed to build the payload preview.'))
  }

  return data as UsageStatisticsPreview
}

/**
 * Reports that the notice reached an administrator, which opens the send gate in a community installation.
 * Rendering is the trigger, so the notice component calls this rather than a button.
 */
export async function recordUsageStatisticsNoticeShown(): Promise<UsageStatisticsSettings> {
  const { data, error, response } = await getClient().POST('/admin/usage-statistics/notice/shown', {})

  if (!response.ok) {
    throw new Error(getApiErrorMessage(error, 'Failed to record the notice.'))
  }

  return data as UsageStatisticsSettings
}

export async function dismissUsageStatisticsNotice(): Promise<UsageStatisticsSettings> {
  const { data, error, response } = await getClient().POST('/admin/usage-statistics/notice/dismiss', {})

  if (!response.ok) {
    throw new Error(getApiErrorMessage(error, 'Failed to dismiss the notice.'))
  }

  return data as UsageStatisticsSettings
}

/** Pretty-prints the payload for display. The wire form is compact; only the whitespace differs. */
export function formatPayloadForDisplay(payload: string): string {
  try {
    return JSON.stringify(JSON.parse(payload), null, 2)
  } catch {
    return payload
  }
}
